sstables_loader.cc

```diff
@@ -343,55 +343,60 @@ struct per_tablet_stream_progress : public stream_progress {
-    auto sstable_it = _sstables.rbegin();
-
-    for (auto tablet_id : _tablet_map.tablet_ids() | std::views::filter([this] (auto tid) { return tablet_in_scope(tid); })) {
-        auto tablet_range = _tablet_map.get_token_range(tablet_id);
-
-        auto sstable_token_range = [] (const sstables::shared_sstable& sst) {
-            return dht::token_range(sst->get_first_decorated_key().token(),
-                                    sst->get_last_decorated_key().token());
-        };
-
-        std::vector<sstables::shared_sstable> sstables_fully_contained;
-        std::vector<sstables::shared_sstable> sstables_partially_contained;
-
-        // sstable is exhausted if its last key is before the current tablet range
-        auto exhausted = [&tablet_range] (const sstables::shared_sstable& sst) {
-            return tablet_range.before(sst->get_last_decorated_key().token(), dht::token_comparator{});
-        };
-        while (sstable_it != _sstables.rend() && exhausted(*sstable_it)) {
-            sstable_it++;
+    auto reversed_sstables = sstables | std::views::reverse;
+    auto sstable_it = reversed_sstables.cbegin();
+
+    for (auto& [tablet_range, sstables_fully_contained, sstables_partially_contained] : tablets_sstables) {
+        // if the sstable's last token is before the tablet's first token, we can skip it
+        auto tablet_first = tablet_range.start().value().value();
+        auto tablet_last = tablet_range.end().value().value();
+        while (sstable_it != reversed_sstables.cend() && (*sstable_it)->get_last_decorated_key().token() <= tablet_first) {
+            ++sstable_it;
         }
 
-        for (auto sst_it = sstable_it; sst_it != _sstables.rend(); sst_it++) {
-            auto sst_token_range = sstable_token_range(*sst_it);
+        for (auto sst_it = sstable_it; sst_it != reversed_sstables.cend(); ++sst_it) {
+            const auto& sst = *sst_it;
+            auto sst_first = sst->get_first_decorated_key().token();
+            auto sst_last = sst->get_last_decorated_key().token();
 
-            // sstables are sorted by first key, so should skip this SSTable since it
-            // doesn't overlap with the current tablet range.
-            if (!tablet_range.overlaps(sst_token_range, dht::token_comparator{})) {
-                // If the start of the next SSTable's token range lies beyond the current tablet's token
-                // range, we can safely conclude that no more relevant SSTables remain for this tablet.
-                if (tablet_range.after(sst_token_range.start()->value(), dht::token_comparator{})) {
-                    break;
-                }
+            // SSTable entirely after tablet -> no further SSTables (larger keys) can overlap
+            if (tablet_last < sst_first) {
+                break;
+            }
+            // SSTable entirely before tablet -> skip and continue scanning later (larger keys)
+            if (sst_last <= tablet_first) {
```

Comment:
don't we lose the sstables here that step right on the boundary of the tablet? I'd imagine `sst_last < tablet_first` is the right one here.